.github/contributing.md

```diff
@@ -0,0 +1,245 @@
+    <br>
+
+    __PHP__
+
+      Formatter: PHP-CS-Fixer
+      Linter: Psalm
+
+    __JS__
+
+      Formatter: Prettier
+      Linter: ESLint
+
+
+<br>
+
+
+- Testing is the cornerstone of producing operative code.
+
+  Every commit to the `master` branch must be tested before merge.
+
+  Include the appropriate test cases for your implemented code.
+
+  This might include:
+
+  - Manual flow tests
+  - Unit tests
+  - MLT(integration) tests
+  - E2E tests
+
+  __Manual Flow Tests__: These are tests performed by contributor by physically walking through the manual process you are working upon to ensure no edge cases and breaking points are present.
+
+  __Unit Tests__: These are tests that focus on individual components or functions in isolation. They are typically low-level tests.
+
+  __Mid-Level Tests (Integration Tests)__: These tests focus on the integration between different components or modules. They ensure that these components work together as expected. These could be considered mid-level tests.
+
+  __End-to-End Tests__: These tests simulate real-world scenarios and interactions, typically spanning multiple components or even the entire application. They can be considered higher-level tests.
+
+
+<br>
+
+- Documentation is the last consideration that should be taken very seriously before committing code changes to the codebase.
+
+  - All types and fns where necessary should be properly documented.
+  Inherently, these comments need to follow `Docblock` or `Psalm` comment standards.
+
+  - No unnecessary comments or `@` tags  should be visible
+
+  - Only avail of docblock syntax to further document types when language doesn't allow concise type declarations.
+
+
+<br>
+
+
+## PR creation and management
+
+To add code to our codebase after following all above steps, you *must* create a new PR for your
+changes.
+
+The destination branch normally is `dev`, ensure this is the case.
+
+Follow and fill in the `PULL_REQUEST_TEMPLATE` present.<br>
+Link corresponding PR to the issue at hand by adding `Closes #{ISSUE-ID}` somewhere in `Description, Changes` section to PR body or add it manually.
+
+> NOTE: Manual option is only available after PR creation from `Development` sidebar option.
+
+If unsure about whether PR is ready for review, create a draft PR.<br>
+Update PR to be full PR and ask for reviewers when ready.
+
+If changes are requested,
+1. Ask for clarification on changes necessary, if needed.
+2. Don't be afraid to state your point of view on proposed changes.
+3. Debate respectfully, acknowledge the other contributors point of view
+4. Finalize decision and solution to identify whether changes are actually necessary
+- If yes, make changes locally. <br>Follow <span style="color: #5C9340;">## Pre-committing to Github</span> guidelines defined above and push changes to Github
+5. Request re-review from contributors once ready again.
+
+> NOTE: Every PR __must__ require at least 2 approvals before merge-ready.
+
```

Comment:
I think this is not so set in stone, especially now, and it's not really a concern of the person submitting the pr

Reply:
I think our case is unique in that I have less experience than both of ye and ye both know Haskell. 
So, in this context I won't be asked for reviews on anything Haskell related but if there are two reviewers who are well versed in the topic/language, it is my opinion this should become the standard. 

The reasons are: 

- two eyes are better than one 
- topic experience outweighs the other reviewer
- diverse skillsets, viewpoints
- various coding styles
- knowledge sharing and learning opportunities 
- reduced biases and preferences
- faster Iterations and reduced technical debt
- backup for busy reviewer
- etc...

I think it should maybe be reframed to say 

```
Every PR must request 2 reviewers.  

Every PR requires 2 approvals unless it falls under one of the following scenarios: 

1. Expedited circumstances - If a small change is urgent and the risk is low, a single approval might be sufficient to move forward

2. Weighted approvals - If the code change falls under a domain of designated "owners" or subject matter experts, their single approval might carry more weight than others who aren't as well knowledgeable.

3. Thoroughly documented changes - If the change is well-documented and the context is clearly explained in the PR description, a single reviewer might be able to understand and assess the change effectively.

4. Team Consensus: If the changes are discussed and identified to conform to a scenario when a single approval is acceptable and when two approvals are unnecessary


In any of these scenarios documented, a single approval is sufficient.
 
```

Reply:
I guess this is really context dependent.

I would not state it in such imperative terms, but more something like: `It is advisable to request a review for a PR to everyone who could provide relevant contribution or has stakes in the code`. Then I would say it's up to every single reviewer to decide whether they want their review to be required, optional or they just want to trust other reviewers.

In general, I think it would be better to state suggestions and guidelines more than rules, since it is more adaptable to how the organization grows

Reply:
Yes, this is a very opinionated issue. 
Here is a forum discussion on the matter, very interesting some of the points put forward 
- https://www.reddit.com/r/csharp/comments/xfk1k7/how_many_people_should_read_pull_request_in_a/

Reply:
Just as an example, in one of the best teams I worked in, we basically abolished PRs (you could request one if you wanted to, otherwise you would just deploy your code). It depends really much on the context you're working on, so I guess being flexible about it is best

Reply:
Okay yes i agree let's make it less of a formality. 
I was reading multiple cases in the above forum where they have multiple devs reviewing but they all just press approve which is way worse than having one thorough review.